### PR TITLE
Using the LinkPreview API to get the page title

### LINKED ISSUE #41: No title fetching on mobile
I've recently noticed that link paste action doesn't fully work on my Obsidian iOS version. I'm saying fully because whenever I paste a link, it's pasted with Markdown link formatting but there's no title fetching happens. Text part just stays blank. 

I've tried waiting and checking if there might be something else causing an issue but everything seems fine. I'm not sure if it's a mobile limitation or a bug but I'd like to have this plugin working as intended on my phone too.

### LINKED ISSUE #46: YouTube link title not being fetched on mobile
Instead of fetching the link title when pasting a link from YouTube.com it just shows YouTube on obsidian mobile.
For eg:
[YouTube](https://youtu.be/Q93ZF4D6mfc)
Instead of 
[The Problem with F1 Sponsorships - YouTube](https://youtu.be/Q93ZF4D6mfc)
This problem occurs for YouTube links on obsidian mobile on Android 10

### LINKED ISSUE #77: Page title fetcher using backend service (on self-hosted server)
Thank you for this wonderful Obsidian plugin. It would be great if there was an option to parse page titles using an external API service. I have written some draft in Golang, check out: https://github.com/ZhymabekRoman/PageTitleify

### LINKED ISSUE #93: [FEATURE] Support Complex Redirects
**Context**
Maybe not possible for my specific use case (company slack URL) - but I have implemented the https://github.com/Ellpeck/ObsidianCustomFrames work-around to get authenticated sessions for automatic titles, but now I am getting "Redirecting.."

This may not be all that doable for within this plugin (website could be using JS, not easy 403 redirects) - but I'll do a little digging later and see what I can report.

**Problem**
- Some URLS (especially authenticated via CustomFrames plugin) return "redirecting.." title, not destination

**Solution**
- Are follow-redirects enabled?
- If JS is required, a more involved virtual browser would likely be required aka less fun and likely way out of scope
- maybe iframe + `iframe.contentWindow.document.title` could work? don't remember XSS limitations here

Thanks for any work-arounds or additional research, and for a handy plugins 99% of the time ;D

### LINKED ISSUE #114: Feature request: Allow login to fetch titles from private sites
Examples:
- Google
- Facebook

### LINKED ISSUE #116: "Blocked by CORS policy" error in console everytime I paste link (title still fetches though)
I get the errors below in console everytime I paste a link. I do not have internet speed issues. 

Although the page title still fetches, the errors in console are disconcerting and I thought worth posting to understand if there's a bigger issue. 

I've noted that these errors have occured for others in the past but seem to have been resolved with fixes so wondering why its occuring for me.  
 
I've updated all plugins. I've also deleted and reinstalled Local Rest API plugin and reset API key (in case your plugin uses this plugin). 

image

### LINKED ISSUE #118: Issue with Title Rendering for ChatGPT, Miro, and Notion Links
Hello,

First off, I want to express how much I appreciate your plugin; it's become an integral part of my daily workflow in Obsidian, and I genuinely love it! I’m reaching out to seek your guidance on a specific issue I’ve encountered that affects my experience slightly.

I frequently link to three sites: ChatGPT, Miro, and Notion, using your plugin. However, I've noticed that the titles for these pages aren't rendering as expected. Could there be a potential workaround or fix for this issue? I'm eager to continue using your excellent plugin without this little hiccup, and any assistance you can provide would be greatly appreciated.

Thank you for your time and for the fantastic work on this plugin.

Best regards

Juan Salazar

### LINKED ISSUE #119: Fetch Reddit link title
it's a problem with Reddit, the title is always "Reddit - Dive into anything" with auto fetch. 

is there a way to fix it? thank you!

### LINKED ISSUE #124: Pasting Youtube link results only in title "YouTube"
It would be nice if it could fetch the video title and append it to "YouTube -"

I'm on iOS Obsidian 1.6.7 (149), using plugin 1.5.4

### PR DESCRIPTION
**Pull request by ZhymabekRoman**
Should fix #124, probably should fix #118, fixes major security problem - #81, closes #77, closes #46, closes #119, closes #116, should fix #114, should fix #93, should fix #41.

Doesn't affect #122

### CODE REVIEW

**Comment by DylanSpeiser** [comment]
Bumping this, looking forward to the YouTube title fix.

**Comment by wlo2** [comment]
Looking forward for YouTube links fix too.

**Comment by yinsanwen** [comment]
Looking forward

File: main.ts [hunk 1]
```diff
@@ -286,22 +292,56 @@ export default class AutoLinkTitle extends Plugin {
...
-    const shortenedTitle = `${title.slice(0, this.settings.maximumTitleLength)}...`;
+    const shortenedTitle = `${title.slice(
+      0,
+      this.settings.maximumTitleLength
+    )}...`;
     return shortenedTitle;
+  };
+
+  public async fetchUrlTitleViaLinkPreview(url: string): Promise<string> {
+    try {
```
Comment (zolrath) [inline main.ts:303]: Looks great!
If we add a check to ensure the `linkPreviewApiKey` is 32 characters long so we don't hit their API with invalid requests when the user hasn't put in an API key I'd be up for merging this!

File: settings.ts [hunk 2]
```diff
@@ -135,5 +137,15 @@ export class AutoLinkTitleSettingTab extends PluginSettingTab {
...
           })
       );
+
+    new Setting(containerEl)
+      .setName("LinkPreview API Key")
+      .setDesc("API key for the LinkPreview.net service. Get one at https://my.linkpreview.net/access_keys")
+      .addText(text => text
+        .setValue(this.plugin.settings.linkPreviewApiKey || "")
+        .onChange(async (value) => {
+          this.plugin.settings.linkPreviewApiKey = value;
```
Comment (zolrath) [inline settings.ts:147]: Also, lets `value.trim()` to remove any erroneous whitespace from the user copy/pasting their API key.

**Review by zolrath — CHANGES_REQUESTED** [review]
This looks great, definitely an awesome service.
A couple of nitpicks so we don't hit their API unnecessarily but overall looks great!

**Comment by ZhymabekRoman** [comment]
@zolrath Done, waiting for review

**Comment by zolrath** [comment]
Works great, thanks for working on this! 🔥

**Pull request merged**